Transfer data

ramirezfranciscof opened the pull request:
Ok, currently the code has the "complete" features, albeit a bit fragily implemented, but now I need a bit of feedback to proceed to make it more robust. First, an overview of how it works:

One needs to provide:
* A list of copying specifications (source node, relative path in source, relative path in target) inside of a instructions input dict.
* The `retrieve_files` logical inside of the instructions input dict (`True` means gather in remote folder and bring back to local folder node, `False` means just leave in the remote folder).
* A `script_copy` optional keyword in the instructions input dict allows to choose one of the files specified to be copied via a command in a script sent through the queue.
* A `remote_path` optional `metadata.option` that, if specified, will make the remote gathering of files in a specific remote path instead of the default setted by the daemon.
* An input `code` (right now this is required always, we should consider making it optional if no `script_copy` is required).
* Finally, the nodes as an input namespace for all the source nodes in the instructions.

The pre-submit is basically divided into two steps:
* First checks if it needs to copy one of the files via the script, and if so it removes it from the local working copy of that list and set ups the paths in the script.
* Then it puts all the remaining copy instructions in the respective lists (`local_copy_list` or `remote_copy_list` depending on the source node, FolderData or RemoteData, and in the `retrieve_list` if the final goal is a local copy).

Now, a few items that I would like to discuss/define so you can have in mind while checking it out:
1. **Namings:** I tried to improve readability a bit from the draft you saw first, but these are far from definite. For example, in the nefarious line `calcinfo.target_path = self.inputs['metadata']['options']['remote_path']`, I do realize it is not good to change variable names like this without good reason, the problem is that I wasn't really sure on which name to go with and since this would eventually be modifying the execmanager I figured it was worth discussing it instead of just going by instinct.
2. **script_copy limitations:** I think this is currently the most limiting aspect of the functionality (not counting the inability to copy from other remotes directly, which I understand more like a "transport" limitation rather than a "transfer" one). As it is stated, right now the script can only copy one single file (or it could copy many files in a single line, but then I don't think you can specify target paths for all of them, so you will be then limiting functionality in that the way relative path location would be copied...). I don't know if we want to extend this functionality or how to. It also requires the `remote_path` optional metadata because it needs to know the name of the target folder in the pre-submit before the upload (which is when the name of the folder is decided). 
3. **Structural issues:** so, the `retrieve_files` is a mandatory flag in the `instructions_dict`, `script_copy` is an optional specification in the `instructions_dict`, and `remote_path` is an optional `metadata.option`. I think we should think if we could "homogenize" them a bit more (a not-necessarily-optimal example of what I mean: make them all optional keys in the `instructions_dict` with defaults), and if not, consider how to harmonize the structure and the functionality (i.e. I think it is a bit "non-friendly" if for enabling the optional `script_copy` dict key you need to also specify a `metadata.option.remote_path` in the CalcJob, as these are pretty much separated). BTW: I know it might sound weird that I am saying "we should change the structure" when it was I that just built it, but I did it to get a working prototype ASAP with increasing functionality in the way I knew how to do it, which might not necessarily be "the most adequate" now that I see the complete final draft, which will probably use tools or considerations that I might not be familiar with yet.
4. **Copying folders recursively:** Right now I am pretty sure that you can't just put the path of a dir instead of a file and expect all these to still work. I'm not sure how much of a priority this is, but my guess is that even before this we should look into how the transport deals when trying to create folders that are not there to complete paths. If you see the comented lines in the first of the test scripts you will see what I mean and that I already had some problems with this, but this would require modifications of the transport so it is a bit more tricky and not sure if we maybe just add a limited version of this before going into that.

ramirezfranciscof committed:
Add transfer calcjob and remove code requirement

sphuber commented on aiida/calculations/plugins/transfer.py:
```diff
@@ -0,0 +1,179 @@
+            'no indication of what to do in the instruction node {}.\n'
+            ' - If you want to retrieve use `retrieve: True`\n'
+            ' - If you want to just gather use `retrieve: False`\n'
+        )
```
The error message does not actually match the check. It fails if the key `retrieve_files` is not present in the `instructions` dict, but the message tells to set the `retrieve` key

sphuber commented on aiida/calculations/plugins/transfer.py:
```diff
@@ -0,0 +1,179 @@
+        errmsg = (
+            'no indication of what to do in the instruction node {}.\n'
+            ' - If you want to retrieve use `retrieve: True`\n'
+            ' - If you want to just gather use `retrieve: False`\n'
```
Same as above, inconsistent check and error message

sphuber commented on aiida/calculations/plugins/transfer.py:
```diff
@@ -0,0 +1,179 @@
+            ' - If you want to just gather use `retrieve: False`\n'
+        )
+        return errmsg.format(instructions.uuid)
+    object_list = instructions.get_dict()['object_list']
```
find the name `object_list` a bit weird when really it contains the instruction set

ramirezfranciscof replied:
Yeah, I know, but I think "instructions" feels more generic and that it include things such as the `retrieve_files` flag which is not part of this list. So the list contains information of each file you want to copy (in which node to find it, relative path within the node, new name/path for the file in the target node) but it does not say where to copy (`FolderData` or `RemoteData`) or how to copy (using the lists or through the script). Would `file_list` be better?...

sphuber commented on aiida/calculations/plugins/transfer.py:
```diff
@@ -0,0 +1,179 @@
+
+    source_uuids = [node.pk for node in source_nodes.values()]
+    source_pkids = [node.uuid for node in source_nodes.values()]
+    instruct_ids = [src_node_id for src_node_id, src_locpath, tgt_locpath in object_list]
```
I think we should only support using UUIDs here for the source node. If we allow pks, the instructions no longer make sense if the calculation is exported. Anyway, then `local_copy_list` also only supports UUIDs and for the user it is the difference between `node.pk` or `node.uuid` which is negligible.

ramirezfranciscof replied:
Also, I just noticed I am storing the pks in the `source_uuids` and the uuids in the `source_pkids`. I only use these to check if they are in either of those lists so it works the same, but funny blunder...

sphuber commented on aiida/calculations/plugins/transfer.py:
```diff
@@ -0,0 +1,179 @@
+    source_pkids = [node.uuid for node in source_nodes.values()]
+    instruct_ids = [src_node_id for src_node_id, src_locpath, tgt_locpath in object_list]
+
+    for node_id in instruct_ids:
+        its_a_uuid = node_id in source_uuids
+        its_a_pkid = node_id in source_pkids
+        if not (its_a_uuid or its_a_pkid):
+            errmsg = 'node identified by `{}` in instructions has not been passed as an input.'
+            return errmsg.format(node_id)
+
+    for nodename, node in source_nodes.items():  # pylint: disable=unused-variable
+        its_a_uuid = node.uuid in instruct_ids
+        its_a_pkid = node.pk in instruct_ids
+        if not (its_a_uuid or its_a_pkid):
+            errmsg = 'input node provided (pk=`{}`, uuid=`{}`) is not used in instructions.'
+            return errmsg.format(node.pk, node.uuid)
```
When we limit to UUIDs only in the instructions, all of this can be simplified a lot.
```
if set(node.uuid for node in source_nodes.values()) != set(instruction[0] for instruction in object_list):
     return 'Incommensurate node UUIDs for the `source_nodes` and `instructions` input`
```

ramirezfranciscof replied:
So, I simplified it a bit, but I think I still preferr when the error says specifically which id was the problem, so I'm still looping twice instead of set comparing. This should be more useful for the user to debug, no?

sphuber commented on aiida/calculations/plugins/transfer.py:
```diff
@@ -0,0 +1,179 @@
+        errmsg = (
+            'no indication of what to do in the instruction node {}.\n'
+            ' - If you want to retrieve use `retrieve: True`\n'
+            ' - If you want to just gather use `retrieve: False`\n'
```
I find the term `gather` as the opposite of `retrieve` a bit confusing. Really this switch means whether to retrieve files to local data or to copy files to remote data.

The fact that copying to remote data, can actually mean to create a `RemoteData` on localhost, is a common quirk of the AiiDA language and should not worry us here.

ramirezfranciscof replied:
So, you mean something like this?

```
' - If you want to copy the files to a FolderData, set `retrieve_files: True`\n'
' - If you want to copy the files to a RemoteData, set `retrieve_files: False`\n'
```

sphuber replied:
I don't think we need to literally say `FolderData` and `RemoteData`. Why not say:
> To store the files in the repository `set retrieve_files=True`, to copy them to the specified folder on the remote computer, set it to `False`

sphuber commented on aiida/calculations/plugins/transfer.py:
```diff
@@ -0,0 +1,179 @@
+        spec.outputs[cls.link_label_retrieved].required = False
+
+        # Resources metadata should only be required if going through a queue
+        spec.inputs['metadata']['options']['resources'].required = False
```
No need to set `required = False` if you specify a default

ramirezfranciscof replied:
Ok, I just commented all these sets to `False` since these are anyways needed for the execmanager not to fail, so this way I am just removing checks but still having to provide the inputs which made no sense...

sphuber commented on aiida/calculations/plugins/transfer.py:
```diff
@@ -0,0 +1,179 @@
+            'num_machines': 1,
+            'tot_num_mpiprocs': 1,
+        }
+
```
as discussed before, we should not actually set such a default as this does not make sense for all schedulers. This is something we should address in the base `CalcJob` class first

sphuber commented on aiida/calculations/plugins/transfer.py:
```diff
@@ -0,0 +1,179 @@
+        calcinfo = CalcInfo()
+        # This one didn't work, so I had to change the accepted type of the input
+        # and assign None as the default...which is the correct way?
+        #calcinfo.target_path = self.inputs.get('metadata.options.remote_path', None)
```
`self.inputs.get('metadata.options.remote_path', None)` does not work because `get` is a normal Python method, but allowing `metadata.options.remote_path` to mean nested dictionaries does not exist. The paradigm of having periods indicate a subspace is something we just implemented in AiiDA for certain methods surrounding the PortNamespaces used for process inputs and outputs.

If you want to do this in normal Python you would have to do:
```
calcinfo.target_path = self.inputs.get('metadata', {}).get('options', {}).get('remote_path', None)
```

ramirezfranciscof replied:
Ah, I see, I thougth the `get` method was overloaded in the `PortNamespaces` to work with the dots. I am not particullarly interested in using normal python, I just wanted a way to access the attribute in a way that even if it was not there I could get a default. Is there a method to do this or do I have to do it like I did with setting a default for the input directly?

sphuber commented on aiida/calculations/plugins/transfer.py:
```diff
@@ -0,0 +1,179 @@
+        # This one didn't work, so I had to change the accepted type of the input
+        # and assign None as the default...which is the correct way?
+        #calcinfo.target_path = self.inputs.get('metadata.options.remote_path', None)
+        calcinfo.target_path = self.inputs['metadata']['options']['remote_path']
```
I think it would still be better to not set `None` as a default and just have it required `False` and then in here just set `None` if the key does not exist. This way, if the key is not used, the node will not have a useless `remote_path: None` attribute.

ramirezfranciscof replied:
> and then in here just set `None` if the key does not exist.

Ok, I see. What would be the best way to do this check with AiiDA tools? You explained before how `self.inputs.get('metadata.options.remote_path', None)` doesn't work, but is there a replacement for this?

(Btw: I thought the `get` method was adapted in the class to work with the dots, I wasn't particullarly interested in working in normal Python, specially if it becomes that much convoluted)

sphuber replied:
How to do it in normal Python I explained in another comment: `self.inputs.get('metadata', {}).get('options', {}).get('remote_path', None)`

sphuber commented on aiida/calculations/plugins/transfer.py:
```diff
@@ -0,0 +1,179 @@
+        # codeinfo.stdout_name = "salidita.out"
+        # codeinfo.stdout_name = self.metadata.options.output_filename
+
+        script_copy = self.inputs['instructions'].get_attribute('script_copy', None)
```
as soon as you are using more attributes of `self.inputs['instructions']` it is probably more efficient to get them in one go `instructions = self.inputs['instructions'].attributes` to prevent multiple database hits.

sphuber commented on aiida/calculations/plugins/transfer.py:
```diff
@@ -0,0 +1,179 @@
+        # codeinfo.stdout_name = self.metadata.options.output_filename
+
+        script_copy = self.inputs['instructions'].get_attribute('script_copy', None)
+        if script_copy:
+            # SHOULD CHECK THAT
+            # (1) NODE IS IN LIST
+            # (2) NODE IS FROM REMOTE
+            # (3) A REMOTE PATH HAS BEEN PROVIDED
```
Why not put this validation in `validate_transfer_inputs`?

ramirezfranciscof replied:
Ah, yes, good point. Anyways, the checks are not there, just the comments, so I will add to it that it eventually should be checked in `validate_transfer_inputs` and maybe wait until we discuss a bit further how to implement this `script_copy` before doing the checks.

sphuber commented on aiida/calculations/plugins/transfer.py:
```diff
@@ -0,0 +1,179 @@
+            instruction = object_list.pop(script_copy - 1)
+            source_node_id = instruction[0]
+            source_relpath = instruction[1]
+            target_relpath = instruction[2]
```
```suggestion
            source_node_id, source_relpath, target_relpath = object_list.pop(script_copy - 1)
```

sphuber commented on aiida/calculations/plugins/transfer.py:
```diff
@@ -0,0 +1,179 @@
+            # (1) NODE IS IN LIST
+            # (2) NODE IS FROM REMOTE
+            # (3) A REMOTE PATH HAS BEEN PROVIDED
+            instruction = object_list.pop(script_copy - 1)
```
This is very confusing to me, so `script_copy` is an integer? That refers to the index of the instruction in the `instructions.object_list` list? Seems like a very weird syntax to me. I realize that you notice that this limits this functionality to only be able to copy a single file through the queue, but even with those restrictions I would not use this syntax

ramirezfranciscof replied:
By "syntax" you mean the name of the variable or the whole concept of passing the position in the list as a way to identify it? Assuming is the latter, I guess I could have it be a single tupple like those in the `instructions.object_list`...is that better? Still not sure about the name in that case...`script_copy_file`?

sphuber commented on aiida/calculations/plugins/transfer.py:
```diff
@@ -0,0 +1,179 @@
+        # it here. I also have some commented sections that might be useful later to pass
+        # parameters
+        codeinfo = CodeInfo()
+        codeinfo.code_uuid = self.inputs['code'].uuid
```
this will crash if no `code` is specified in the inputs. By the way, what kind of `Code` is it expected to be anyway?

ramirezfranciscof replied:
What do yo mean by "kind" of code? What are the possible "kinds"?

sphuber replied:
Well, the `PwCalculation` is made for Quantum ESPRESSO's `pw.x` and so in AiiDA the `Code` has `input_plugin = quantumespresso.pw`. What kind of binary is this `Code` supposed to be run on the remote? Are there any limitations? Should there be?

sphuber commented on aiida/calculations/plugins/transfer.py:
```diff
@@ -0,0 +1,179 @@
+            source_node = orm.load_node(source_node_id)
+            source_path = os.path.join(source_node.get_remote_path(), source_relpath)
+            target_path = os.path.join(calcinfo.target_path, source_relpath)
+            if not os.path.isabs(calcinfo.target_path):
+                code_node = orm.load_node(self.inputs['code'].uuid)
+                computer_path = code_node.computer.get_workdir()
+                target_path = os.path.join(computer_path, target_path)
```
I would let the making absolute of this path up to the `upload_calculation` method of the manager.

ramirezfranciscof replied:
But the problem is that I need to pass the absolute paths as the `codeinfo.cmdline_params` now as instructions for the "cp/rsync" command inside the script.

ramirezfranciscof replied:
BTW: Notice the `code_node = orm.load_node(self.inputs['code'].uuid)`. If that is not art, I don't know what is...

sphuber commented on aiida/calculations/plugins/transfer.py:
```diff
@@ -0,0 +1,179 @@
+
+        for source_node_id, source_relpath, target_relpath in object_list:
+
+            source_node = orm.load_node(source_node_id)
```
no need to load the node as they should already be in `self.inputs['source_nodes']`

sphuber commented on aiida/calculations/plugins/transfer.py:
```diff
@@ -0,0 +1,179 @@
+            source_node_id = instruction[0]
+            source_relpath = instruction[1]
+            target_relpath = instruction[2]
+            source_node = orm.load_node(source_node_id)
```
Probably want to prevent loading the node again as it is already in `self.inputs['source_nodes']`. You can make a useful mapping once that you can reuse to get node instance given its uuid:
```
node_mapping = {node.uuid: node for node in self.inputs['source_nodes']}
```

sphuber commented on aiida/calculations/plugins/transfer.py:
```diff
@@ -0,0 +1,179 @@
+
+        calcinfo.codes_info = [codeinfo]
+
+        for source_node_id, source_relpath, target_relpath in object_list:
+
+            source_node = orm.load_node(source_node_id)
+
+            if isinstance(source_node, orm.FolderData):
+                source_uuid = source_node.uuid
+                source_path = source_relpath
+                target_path = target_relpath
+                calcinfo.local_copy_list.append((source_uuid, source_path, target_path))
```
This could probably be:
```
for instruction in object_list:

    source_node = node_mapping[instruction[0])
    if isinstance(source_node, orm.FolderData):
        calcinfo.local_copy_list.append(instruction)
```
If you feel that `instruction[0]` is not very readable, I would create a namedtuple and transform the instructions in those tuples once and then you can use `instruction.source_uuid` for readability.

sphuber commented on aiida/calculations/plugins/transfer.py:
```diff
@@ -0,0 +1,179 @@
+                target_path = target_relpath
+                calcinfo.remote_copy_list.append((source_uuid, source_path, target_path))
+
+            else:
+                raise ValueError(
+                    'type of source node id={} is not allowed ({})'.format(source_node_id, source_node.node_type)
+                )
```
This should never happen due to the validation of the process spec. So this should probably be a `RuntimeError` or something

sphuber commented on aiida/engine/daemon/execmanager.py:
```diff
@@ -41,6 +41,9 @@ def upload_calculation(node, transport, calc_info, folder, inputs=None, dry_run=
     from tempfile import NamedTemporaryFile
     from aiida.orm import load_node, Code, RemoteData
 
+    debug_printing = False
+    if debug_printing: print('Je commence')
```
I realize that this is just for your debugging purposes, but you might want to just the built in logging for this. Simply turn these in `logger.debug('msg')` statements and set `verdi config logging.aiida_loglevel DEBUG` and presto

ramirezfranciscof replied:
So, what would be the advantage of this? I understand that in the production version having a log for the user is better than just messages popping out everywhere and having the code full of dirty if prints, but if these messages are going to be taken out is there any advantage of using the logger?

sphuber commented on aiida/engine/daemon/execmanager.py:
```diff
@@ -75,55 +81,78 @@ def upload_calculation(node, transport, calc_info, folder, inputs=None, dry_run=
-            logger.debug(
-                '[submission of calculation {}] Unable to chdir in {}, trying to create it'.format(
-                    node.pk, remote_working_directory))
+        if calc_info.target_path:
+            pathlist = os.path.normpath(calc_info.target_path).split(os.sep)
+        else:
+            pathlist = [None]
+
+        # If it is an absolute path it will start checking dirs from the root
+        if pathlist[0] == '':
+            transport.chdir('/')
+            pathlist.pop(0)
+
+        # Else (relative path or no target path provided), it will start from
+        # the computer working directory
+        else:
```
This should be changed by just defining `remote_working_directory`:
```
if calc_info.target_path:
    if os.path.isabs(calc_info.target_path):
        remote_working_directory = calc_info.target_path
    else:
        remote_working_directory = os.path.join(remote_working_directory, calc_info.target_path)
```

ramirezfranciscof replied:
Mmm ok, but the existing logic in the execmanager has hardcoded the logic of the 2:2:32 sharding, do you want me to fully replace this for something more versatile? If that is the case, do you think that the "path as list" approach is good? So I mean:

```
(1) set remote_working_directory to the default sharding or to calc_info.target_path
(2) pathlist = os.path.normpath(remote_working_directory).split(os.sep)
(3) Go element by element making sure the dir exists until the last one
(4) Last one should be new or else apply existing procedure of moving this to REMOTE_WORK_DIRECTORY_LOST_FOUND
```

ramirezfranciscof replied:
Just a comment for the future (as we decided not to implement the `remote_path` yet): the `remote_working_directory` is the base location where the remote directories will be created according to the aiida sharding of the node uuid, not the final path (as the one passed through `target_path`). So these variables may need some re-organization, I can't just replace `remote_working_directory`. Also relevant for this is that, although `remote_working_directory` is almost exclusively used in lines 72 - 93 (where it just checks it exists and chdirs to it), it is also used one single time later (line 113) to create the `LOST_FOUND` directory. I should open an issue with all these things before closing this PR.

sphuber commented on aiida/engine/daemon/execmanager.py:
```diff
@@ -75,55 +81,78 @@ def upload_calculation(node, transport, calc_info, folder, inputs=None, dry_run=
-        transport.chdir(calc_info.uuid[:2])
-        transport.mkdir(calc_info.uuid[2:4], ignore_existing=True)
-        transport.chdir(calc_info.uuid[2:4])
+            except IOError:
+                logger.debug(
+                    '[submission of calculation {}] Unable to chdir in {}, trying to create it'.format(
+                        node.pk, remote_working_directory))
+                try:
+                    transport.makedirs(remote_working_directory)
+                    transport.chdir(remote_working_directory)
+                except EnvironmentError as exc:
+                    raise exceptions.ConfigurationError(
+                        '[submission of calculation {}] '
+                        'Unable to create the remote directory {} on '
+                        "computer '{}': {}".format(
+                            node.pk, remote_working_directory, computer.name, exc))
+
+        if calc_info.target_path:
+            for pathpart in pathlist:
+                try:
+                    transport.chdir(pathpart)
+                except IOError:
+                    transport.mkdir(pathpart)
+                    transport.chdir(pathpart)
```
All of this logic should be simplified and improved a lot. All that this is doing is given an absolute path, try to create the path and chdir into it. This should be independent of whether `calc_info.target_path` is actually defined, so all this different logic can be gotten rid off

ramirezfranciscof replied:
Mmm ok, but the existing logic in the execmanager has hardcoded the logic of the 2:2:32 sharding, do you want me to fully replace this for something more versatile? If that is the case, do you think that the "path as list" approach is good? So I mean:

1. set remote_working_directory to the default sharding or to calc_info.target_path
2. ```pathlist = os.path.normpath(remote_working_directory).split(os.sep)```
3. Go element by element making sure the dir exists until the last one
4. Last one should be new or else apply existing procedure of moving this to REMOTE_WORK_DIRECTORY_LOST_FOUND

sphuber replied:
I would determine the dirname, so the full path minus the last section, and use `os.makedirs(dirname, exist_ok=True)` which will not raise since [Python 3.4](https://docs.python.org/3/library/os.html#os.makedirs). Then you just create the leaf, which if it already exists, you enter the procedure for backing up. The only part that is different for your application is determination of `dirname`. So there you have a single switch and then all the rest becomes the same code

ramirezfranciscof replied:
This would have to be implemented **through the transport**, right? Because I can't directly use `os` to do stuff on the "remote" machine.

sphuber reviewed (changes requested):
Thanks @ramirezfranciscof , great to have a working concept. As you can see, there are quite a few things to improve still. Happy to discuss this in person which is probably easier than doing everything on GIthub

ramirezfranciscof committed:
Remove requirement of passing resources

Did so by adding defaults in the calcjob implementation for the main
required keywords: num_machines and num_mpiprocs_per_machine.

A next step would be looking if these can be changed so as to default to
an empty dict instead, but these keywords are being checked deep into
the code (I could change the behaviour of the schedulers plugin, but it
is then also passed to mpi_args in the CalcJob in a very "delocallized"
way).

ramirezfranciscof committed:
Split file list into local, remote, symlink

ramirezfranciscof committed:
Add docstring documentation to CalcJob

ramirezfranciscof commented:
**UPDATE ON THE TRANSFER CALCJOB**

1. The "ab-initio" instructions on what this calcjob is and how it works have been included in the docstring, so you can view the there and this should be understandable hopefully. Here I will just list the changes and tests I made.

2. I removed the functionalities of naming your `RemoteData` folder and of copying by sending a script with a copy command to the queue. Curently it will only copy files using the lists set up in the general `CalcJob` and recognized by the engine.

3. The code now uses the label in the `source_node` dict instead of the uuid to identfy nodes in the instructions. Also, instead of one list there are 3: `local_files`, `remote_files` and `symlink_files` (I tested that using symlings [works just fine](https://github.com/ramirezfranciscof/aiida-core/issues/7)). Since dividing the list into 2 was already in the cards, I thought doing it in 3 was better than having to deal with the 3-tupples / 4 tupple duality...

4. I removed the requirement of providing a code and added the requirement to provide a computer: after some testing I discovered that the engine is already able to deal with this by passing an empty list in `calcinfo.codes`. In the `CalcJob` node, I decided to set the default as `Code=None` instead of just popping out the code from the specs because it might be better for backwards consistency, both for the machines (example: implemented queries might be able to handle “unidentified” codes but may break if the “code” is not even an entry) and the humans (since until now the code has always been a strong requirement, seeing "Code: None" could be less confusing than seeing no code at all).

5. I also removed the requirement of providing `metadata.options.resources` in the same way that it was implemented in the new `Arithmetic.Add`, by setting the default to be:
```
spec.inputs['metadata']['options']['resources'].default = { 'num_machines': 1, 'num_mpiprocs_per_machine': 1 }
```
I did try to go into the way these are being used and even tried to re-arrange the procedures inside of the scheduler (see [this other PR](https://github.com/ramirezfranciscof/aiida-core/pull/8) in case you think this work it is worth including in AiiDA-core) so that there was no need to set these up, but then the problems started appearing in the parent `CalcJob` class. I mean, I can also try to change the structure there (since the checks and the application are made in different places, I would either have to hack it or re-arrange this structure as well), but I just wanted to first check with you to see if it was worth it keep going this way or if we had other priorities for this (specially since, even when not using mpi, these defaults may be a better description of what is happening than an empty dict or no dict at all).

6. The engine (more specifically, I suspect this is a shortcoming of the transport) [can't create folders](https://github.com/ramirezfranciscof/aiida-core/issues/5) when used as part of the key of a target file.

7. The engine (more specifically, I suspect this is a shortcoming of the local repository methods) [can't recursively copy folders](https://github.com/ramirezfranciscof/aiida-core/issues/6). I'm not sure if we want to focus on fixing this with all the repo redesign going on.

8. We never said anything about doing this yet, but since I think it is a relevant limitation (at least in terms that it might be hard for the user to understand why this is limited), just wanted to keep tabs on the fact that this can only copy files between the local computer and a single remote computer (so you can't copy files from remote to remote).

Well, I think this is all we discussed the last time...(btw: damnit, I swear to thor that I try to keep these explanations as short as possible...)

sphuber commented on aiida/calculations/transfer.py:
```diff
@@ -0,0 +1,211 @@
+
+        # The transfer just needs a computer, the code are resources are set here
+        spec.inputs['code'].default = None
+        spec.inputs['code'].valid_type = (orm.Code, type(None))
```
```suggestion
        spec.inputs['code'].required = False
```
This has the same effect, as it will allow `None` and is nicer than doing `type(None)`

ramirezfranciscof replied:
So, a couple of questions:
1. I seem to remember than settig a default (previous line, 121) already implied that `required=False`, right?
2. Setting the required flag to False automatically make `None` a valid option, even if the input has a default different than `None`?

sphuber replied:
Ah I see now, yes, setting `default=True` sets `required=False` and so values of `None` will be accepted. So you can undo the change I did and simply remove `type(None)`. But, since the base class already defines the type, just setting `required=False` is enough and you don't need to set the `default` nor change the `valid_type`

sphuber commented on aiida/calculations/transfer.py:
```diff
@@ -0,0 +1,211 @@
+        spec.inputs['code'].default = None
+        spec.inputs['code'].valid_type = (orm.Code, type(None))
+        spec.inputs['metadata']['computer'].required = True
+        spec.inputs['metadata']['options']['resources'].default = {
+            'num_machines': 1,
+            'num_mpiprocs_per_machine': 1,
+        }
```
We should look to try and fix this in `aiida-core`. I will see if I can find the find to have a look such that it is no longer required by default for all calculations and all schedulers, as it shouldn't

ramirezfranciscof replied:
Well, you can check [the other PR](https://github.com/ramirezfranciscof/aiida-core/pull/8) in my fork where I already changed a bit the structure of the checks performed by the scheduler on the resources. Previously it had multiple raises in different conditional branches, but now it is all unified in one so it should be as easy as setting a flag that disables that single raise (currently `requires_resources`) or removing that raise altogether.

On the other hand, at the end of the process I still had to do some setting:
```
        if self.num_machines is None:
            self.num_machines = 1
        (...)
        if self.num_mpiprocs_per_machine is None:
            self.num_mpiprocs_per_machine = 1 
```
Because this is then used again in `presubmit` of the `CalcJob`. Line 382 ALWAYS tries to load the total number of processors:
```
subst_dict = {'tot_num_mpiprocs': job_tmpl.job_resource.get_tot_num_mpiprocs()}
for key, value in job_tmpl.job_resource.items():
      subst_dict[key] = value
mpi_args = [arg.format(**subst_dict) for arg in computer.get_mpirun_command()]
```

Yet this (`mpi_args`) is only used if going with mpi (lines 414-419):
```
if this_withmpi:
    this_argv = (mpi_args + extra_mpirun_params + [this_code.get_execname()] +
                             (code_info.cmdline_params if code_info.cmdline_params is not None else []))
else:
    this_argv = [this_code.get_execname()] + (code_info.cmdline_params
                                                          if code_info.cmdline_params is not None else [])
```

I mean, if this is the last guard (which seems to be the case) it shouldn't be hard to change, I can add the modifications to the branch where I adapted the scheduler, I just wanted to check with you how to you if this is just putting the `subst_dict` parsing inside the "if mpi" clause or if perhaps we need to discuss something regarding how we conceptualize the meaning of this parameters (i.e. it could be say that even when not using mpi, whatever command you run is using at least some resources on the machine and thus you need at least one processor of at least one computer...).

sphuber commented on aiida/calculations/transfer.py:
```diff
@@ -0,0 +1,211 @@
+            'num_mpiprocs_per_machine': 1,
+        }
+
+        # Retrieved node should only be generated if actually retrieving files
+        #spec.outputs[cls.link_label_retrieved].required = False
+        #spec.output(cls.link_label_retrieved).required = False
```
Did we decide that the `retrieved` output should *always* be there, even if `retrieve_files=False`?

ramirezfranciscof replied:
Noup; I didn't change this because I seem to remember we briefly mentioned this in our last meeting but went on with other issues and didn't decide what to finally do with this. Or if we did I forgot / didn't write it down, sorry =S.

sphuber commented on aiida/calculations/transfer.py:
```diff
@@ -0,0 +1,211 @@
+        spec.inputs.validator = validate_transfer_inputs
+
+    def prepare_for_submission(self, folder):
+        source_nodes = self.inputs['source_nodes']
```
```suggestion
        source_nodes = self.inputs.source_nodes
```

ramirezfranciscof replied:
So, I'm a bit confused. After the previous review with my attempt of trying to use defaults when getting this dicts and the python normal way of doing stuff, I got the impresion that your final conclusion was that you were regretting the widespread use of this way of accessing the inputs since it lead to misunderstanding of its limitations. Why now you prefer that I access it this way, is there any advantage?

sphuber commented on aiida/calculations/transfer.py:
```diff
@@ -0,0 +1,211 @@
+
+    def prepare_for_submission(self, folder):
+        source_nodes = self.inputs['source_nodes']
+        instructions = self.inputs['instructions'].get_dict()
```
```suggestion
        instructions = self.inputs.instructions.get_dict()
```

sphuber commented on aiida/calculations/transfer.py:
```diff
@@ -0,0 +1,211 @@
+        for source_label, source_relpath, target_relpath in local_files:
+
+            source_node = source_nodes[source_label]
+            if not isinstance(source_node, orm.FolderData):
+                raise RuntimeError(
+                    'type of node {} for file {} in `local_files` list'
+                    'must be orm.FolderData instead of {}'.format(
+                        source_label,
+                        source_relpath,
+                        source_node.node_type,
+                    )
+                )
```
Why not have this validation already in the spec validator?

ramirezfranciscof replied:
I did think about this, but I thought that since here I'm cycling through the files here and not in the validation it was better this way than doing double the cycles. I guess that this could be a premature optimization, and also I could maybe explicitly do the conversions of lists into sets and do this check there instead of using the list compression and technically keep the same number of cycles...

sphuber replied:
Good point, but in this case this performance hit is acceptable, because know, if the value is incorrect, you will except in `prepare_for_submission`. At that point the node will already have been stored in the database. I think it is preferable to have slight inefficiency in order to prevent unnecessary excepted nodes in the database

sphuber commented on aiida/calculations/transfer.py:
```diff
@@ -0,0 +1,211 @@
+
+            source_uuid = source_node.uuid
+            source_path = source_relpath
+            target_path = target_relpath
```
why this temporary variable assignment? Don't seem to do much with it and it also doesn't really make things much clearer

sphuber commented on aiida/calculations/transfer.py:
```diff
@@ -0,0 +1,211 @@
+        for source_label, source_relpath, target_relpath in remote_files:
+
+            source_node = source_nodes[source_label]
+            if not isinstance(source_node, orm.RemoteData):
+                raise RuntimeError(
+                    'type of node {} for file {} in `remote_files` list'
+                    'must be orm.RemoteData instead of {}'.format(
+                        source_label,
+                        source_relpath,
+                        source_node.node_type,
+                    )
+                )
```
Same here: why not in spec validator?

sphuber commented on aiida/calculations/transfer.py:
```diff
@@ -0,0 +1,211 @@
+    source_nodes = inputs['source_nodes']
+    instructions = inputs['instructions']
+
+    if not 'retrieve_files' in instructions.get_dict():
```
you call `instructions.get_dict()` again on line 31. Better call it once and assign to variable

sphuber replied:
I would put the validation of the `instructions` that does not depend on `source_nodes` in its own validator and set that directly on the `spec.inputs.instructions` port. The validators of ports within a port namespace are always called before the validator of the namespace itself. That means you can safely assume that if you get to the `inputs` validator, the `instructions` validator has passed and you can assume all the required keys are there.

sphuber commented on aiida/calculations/transfer.py:
```diff
@@ -0,0 +1,211 @@
+    remote_files = instructions_dict.get('remote_files', None)
+    symlink_files = instructions_dict.get('symlink_files', None)
+
+    if (local_files is None) and (remote_files is None) and (symlink_files is None):
```
```suggestion
    if not any([local_files, remote_files, symlink_files]):
```

sphuber commented on aiida/calculations/transfer.py:
```diff
@@ -0,0 +1,211 @@
+    The `instructions` dict must also contain at least one list with specifications of which files
+    to copy and from where. All these lists take tupples of 3 that have the following format:
+
+        [ ..., ( source_node_label, path_to_file_in_source, path_to_file_in_target), ... ]
```
Should probably make very clear that `source_node_label` corresponds to the key used in the `source_nodes` namespace

sphuber commented on aiida/calculations/transfer.py:
```diff
@@ -0,0 +1,211 @@
+        errmsg = (
+            'no indication of which files to copy were found in the instruction node {}.\n'
+            'Please include at least one of `local_files`, `remote_files`, or `symlink_files`.\n'
+            'These should be lists containing tupples following the pattern:\n'
```
```suggestion
            'These should be lists containing tuples following the pattern:\n'
```

sphuber commented on aiida/calculations/transfer.py:
```diff
@@ -0,0 +1,211 @@
+            'no indication of which files to copy were found in the instruction node {}.\n'
+            'Please include at least one of `local_files`, `remote_files`, or `symlink_files`.\n'
+            'These should be lists containing tupples following the pattern:\n'
+            '[ ... (source_uuid, source_relpath, target_relpath) ... ] \n'
```
```suggestion
            '[ ... (source_node_key, source_relpath, target_relpath) ... ] \n'
```

sphuber commented on aiida/calculations/transfer.py:
```diff
@@ -0,0 +1,211 @@
+# For further information on the license, see the LICENSE.txt file        #
+# For further information please visit http://www.aiida.net               #
+###########################################################################
+"""Implementation of CalcJobNode to add two numbers for testing and demonstration purposes."""
```
old docstring

sphuber commented on aiida/calculations/transfer.py:
```diff
@@ -0,0 +1,211 @@
+        (2) create a local FolderData and copy there all the files (`retrieve_files = True`)
+
+    The `instructions` dict must also contain at least one list with specifications of which files
+    to copy and from where. All these lists take tupples of 3 that have the following format:
```
```suggestion
    to copy and from where. All these lists take tuples of 3 that have the following format:
```

sphuber reviewed (commented):
Have some comments and suggestions that we can discuss together tomorrow

ramirezfranciscof committed:
Apply PR corrections to the Transfer calcjob.

ramirezfranciscof commented:
Ok, so I am not sure why some of the tests are failing (or why they seem to be duplicated), but I think I applied all required changes. @sphuber you can re-review it whenever you have the time, in the meantime I am going to start setting up aiida-QE and testing the calcjob with it.

ramirezfranciscof committed:
Make paths to put files in transport

ramirezfranciscof commented:
@sphuber I had to "solve" the problem of not being able to create directories with the transport because the restart density needs to be put in the "/out" dir for it to be recognized by QE. I did so in the put methods by adding a mkdir to the path just before that to make sure it exists. Probably the one for `local` should use the `self.makedirs` instead of the `os.makedirs` (like in the `ssh` plugin), but I wanted to make sure to get your feedback on which would be the best way to put this before going into the details of how to put it. Anyways, just letting you know what that last push was about.
